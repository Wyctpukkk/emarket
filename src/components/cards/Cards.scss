.cards {
  &__list {
    display: flex;
    justify-content: space-between;
    flex-wrap: wrap;
    gap: 21px;
  }
}
.card {
  width: 326px;
  height: 490px;
  background: #ffffff;
  box-shadow: 0px 14px 39px -11px rgba(0, 0, 0, 0.06);
  border-radius: 10px;
  padding: 23px;
  &__image-wrapper {
    height: 228px;
    width: 100%;
  }
  &__image {
  }

  &__desc {
    font-weight: 500;
    font-size: 16px;
    line-height: 130%;
    color: #111111;
    margin: 15px 0 20px 0;
  }

  &__title {
    font-size: 16.5px;
    font-weight: 700;
  }

  &__details {
    &_item {
      margin: 5px 0 5px 0;
      span {
        font-weight: 700;
        margin-left: 5px;
      }
    }
  }

  &__payload {
    position: relative;
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 15px;
  }
  &__cost {
    font-weight: 800;
    font-size: 16px;
    line-height: 130%;
    color: #111111;
  }

  &__add {
    margin-left: 15px;
    width: 153px;
    height: 45px;
    background: #ffc85e;
    border-radius: 80px;
    font-weight: 700;
    font-size: 10px;
    line-height: 12px;
    letter-spacing: 0.15em;
    color: #ffffff;
    padding-right: 25px;
    &_icon {
      position: absolute;
      width: 24px;
      height: 17px;
      top: 12px;
      right: 27px;
      background: url(../../assets/cart-white.svg);
    }
  }
}
