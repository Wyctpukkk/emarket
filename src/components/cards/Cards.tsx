import { useEffect, useState } from 'react';
import { useAppSelector, useAppDispatch } from '../../hooks/redux';
import { Link } from 'react-router-dom';
import { Product } from '../../models/IProduct';
import { addCatalogProduct } from '../../store/reducers/cartSlice';
import { Pagination } from '../pagination/Pagination';

const Cards: React.FC = () => {
  const dispatch = useAppDispatch();
  const db: Product[] = useAppSelector((state) => state.dataReducer.database);
  useEffect(() => {
    setItems(db);
  }, [db]);

  const [items, setItems] = useState<Product[]>(db);

  const [currentPage, setCurrentPage] = useState<number>(1);
  const [itemsPerPage] = useState<number>(9);
  const lastItemIndex: number = currentPage * itemsPerPage;
  const firtItemIndex: number = lastItemIndex - itemsPerPage;
  const currentItem: Product[] = items.slice(firtItemIndex, lastItemIndex);
  const paginate = (pageNumber: number) => {
    setCurrentPage(pageNumber);
  };

  const {
    category = 0,
    search = '',
    minPrice = 0,
    maxPrice = 10000,
    brand = '',
  } = useAppSelector((state) => state.sortReducer);

  const sortProperty: string = useAppSelector(
    (state) => state.sortReducer.sort.sortProperty
  );

  const onClickAddCatalogProduct = (obj: Product) => {
    dispatch(addCatalogProduct(obj));
  };

  useEffect(() => {
    if (category === 0) {
      setItems(db);
    } else {
      setItems([
        ...db.filter((obj: Product) => {
          return obj.category.includes(category);
        }),
      ]);
    }
  }, [category, db]);

  useEffect(() => {
    if (search === '') {
      setItems(db);
    } else {
      setItems([
        ...db.filter((obj: Product) => {
          return obj.productedBy.toLowerCase().includes(search.toLowerCase());
        }),
      ]);
    }
  }, [search, db]);

  useEffect(() => {
    setItems([
      ...db.filter((obj: Product) => {
        return obj.price > minPrice && obj.price < maxPrice;
      }),
    ]);
  }, [minPrice, maxPrice, db]);

  useEffect(() => {
    if (brand.length === 0) {
      setItems(db);
    } else {
      setItems([
        ...db.filter((obj: Product) => {
          return brand.includes(obj.brand);
        }),
      ]);
    }
  }, [brand, db]);

  useEffect(() => {
    switch (sortProperty) {
      case 'priceUp':
        setItems([
          ...db.slice().sort((a: Product, b: Product) => {
            if (a.price > b.price) {
              return 1;
            }
            if (a.price < b.price) {
              return -1;
            }
            return 0;
          }),
        ]);
        break;

      case 'priceDown':
        setItems([
          ...db.slice().sort((a: Product, b: Product) => {
            if (a.price > b.price) {
              return -1;
            }
            if (a.price < b.price) {
              return 1;
            }
            return 0;
          }),
        ]);
        break;

      case 'titleDown':
        setItems([
          ...db.slice().sort((a: Product, b: Product) => {
            if (a.name > b.name) {
              return 1;
            }
            if (a.name < b.name) {
              return -1;
            }
            return 0;
          }),
        ]);
        break;

      case 'titleUp':
        setItems([
          ...db.slice().sort((a: Product, b: Product) => {
            if (a.name > b.name) {
              return -1;
            }
            if (a.name < b.name) {
              return 1;
            }
            return 0;
          }),
        ]);
        break;
    }
  }, [sortProperty, db]);

  const categories = useAppSelector((state) => state.dataReducer.categoryArray);

  const showCategories = (arr: number[]) => {
    const names: string[] = [];
    categories.map((obj) => {
      return arr.includes(obj.categoryProperty)
        ? names.push(' ' + obj.first + ' ' + obj.second)
        : '';
    });

    return names.join(',').toLowerCase();
  };

  return (
    <div className="cards">
      {items && (
        <ul className="cards__list">
          {currentItem.map((obj) => {
            return (
              <li key={obj.uid} className="card">
                <div className="card__image-wrapper">
                  <img className="card__image" src={obj.url} alt="product" />
                  <div className="card__type">
                    <div
                      className={`card__type_${
                        obj.type === 'л' ? 'litr' : 'kg'
                      }`}
                    ></div>
                    {obj.typeValue} {obj.type}
                  </div>
                </div>
                <Link to={`item/${obj.uid}`} className="card__desc">
                  <span className="card__title">{obj.name}</span> {obj.desc}
                </Link>
                <ul className="card__details">
                  <li className="card__details_item">
                    Штрихкод:<span>{obj.uid}</span>
                  </li>
                  <li className="card__details_item">
                    Производитель:<span>{obj.productedBy}</span>
                  </li>
                  <li className="card__details_item">
                    Бренд:<span>{obj.brand}</span>
                  </li>
                  <li className="card__details_item">
                    Тип ухода:<span>{showCategories(obj.category)}</span>
                  </li>
                </ul>
                <div className="card__payload">
                  <span className="card__cost">{obj.price} ₸</span>
                  <button
                    onClick={() => {
                      onClickAddCatalogProduct(obj);
                    }}
                    className="card__add"
                  >
                    В КОРЗИНУ
                  </button>
                  <span className="card__add_icon"></span>
                </div>
              </li>
            );
          })}
        </ul>
      )}
      <div className="pagination-block">
        <Pagination
          currentPage={currentPage}
          itemsPerPage={itemsPerPage}
          totalItems={items.length}
          paginate={paginate}
        />
      </div>
      <p className="decription">
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam interdum
        ut justo, vestibulum sagittis iaculis iaculis. Quis mattis vulputate
        feugiat massa vestibulum duis. Faucibus consectetur aliquet sed
        pellentesque consequat consectetur congue mauris venenatis. Nunc elit,
        dignissim sed nulla ullamcorper enim, malesuada.
      </p>
    </div>
  );
};

export { Cards };
