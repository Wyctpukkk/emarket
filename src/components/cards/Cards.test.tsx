import { render, screen } from '@testing-library/react';
import { Cards } from './Cards';
import { MemoryRouter } from 'react-router-dom';
import * as reduxHooks from 'react-redux';

jest.mock('react-redux');

describe('Header', () => {
  test('Создание снапшота', () => {
    jest.spyOn(reduxHooks, 'useSelector').mockReturnValue([
      {
        url: 'https://avatars.mds.yandex.net/get-mpic/4370207/img_id8174150561422744942.jpeg/orig',
        name: 'Neutrale',
        type: 'л',
        typeValue: 0.95,
        uid: 4604049097548,
        category: [9],
        productedBy: 'Neutrale',
        brand: 'Neutrale',
        desc: 'гель для стирки детского белья гипоаллергенный',
        price: 281,
        count: 1,
      },
    ]);

    const view = render(
      <MemoryRouter>
        <Cards />
      </MemoryRouter>
    );
    const cardIsReal = screen.getByText(
      /гель для стирки детского белья гипоаллергенный/i
    );
    expect(cardIsReal).toBeInTheDocument();
    expect(view).toMatchSnapshot();
  });
});
