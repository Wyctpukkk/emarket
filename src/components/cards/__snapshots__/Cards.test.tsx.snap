// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header Создание снапшота 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="cards"
      >
        <ul
          class="cards__list"
        >
          <li
            class="card"
          >
            <div
              class="card__image-wrapper"
            >
              <img
                alt="product"
                class="card__image"
                src="https://avatars.mds.yandex.net/get-mpic/4370207/img_id8174150561422744942.jpeg/orig"
              />
              <div
                class="card__type"
              >
                <div
                  class="card__type_litr"
                />
                0.95
                 
                л
              </div>
            </div>
            <a
              class="card__desc"
              href="/item/4604049097548"
            >
              <span
                class="card__title"
              >
                Neutrale
              </span>
               
              гель для стирки детского белья гипоаллергенный
            </a>
            <ul
              class="card__details"
            >
              <li
                class="card__details_item"
              >
                Штрихкод:
                <span>
                  4604049097548
                </span>
              </li>
              <li
                class="card__details_item"
              >
                Производитель:
                <span>
                  Neutrale
                </span>
              </li>
              <li
                class="card__details_item"
              >
                Бренд:
                <span>
                  Neutrale
                </span>
              </li>
              <li
                class="card__details_item"
              >
                Тип ухода:
                <span />
              </li>
            </ul>
            <div
              class="card__payload"
            >
              <span
                class="card__cost"
              >
                281
                 ₸
              </span>
              <button
                class="card__add"
              >
                В КОРЗИНУ
              </button>
              <span
                class="card__add_icon"
              />
            </div>
          </li>
        </ul>
        <div
          class="pagination-block"
        >
          <div
            class="pagination"
          >
            <button
              class="pagination__prev"
            >
              <img
                alt="prev"
                src="pagination.svg"
              />
            </button>
            <ul
              class="pagination__list"
            >
              <li
                class="pagination__item"
              >
                <button
                  class="pagination__page active"
                >
                  1
                </button>
              </li>
            </ul>
            <button
              class="pagination__next"
            >
              <img
                alt="next"
                src="pagination.svg"
              />
            </button>
          </div>
        </div>
        <p
          class="decription"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam interdum ut justo, vestibulum sagittis iaculis iaculis. Quis mattis vulputate feugiat massa vestibulum duis. Faucibus consectetur aliquet sed pellentesque consequat consectetur congue mauris venenatis. Nunc elit, dignissim sed nulla ullamcorper enim, malesuada.
        </p>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="cards"
    >
      <ul
        class="cards__list"
      >
        <li
          class="card"
        >
          <div
            class="card__image-wrapper"
          >
            <img
              alt="product"
              class="card__image"
              src="https://avatars.mds.yandex.net/get-mpic/4370207/img_id8174150561422744942.jpeg/orig"
            />
            <div
              class="card__type"
            >
              <div
                class="card__type_litr"
              />
              0.95
               
              л
            </div>
          </div>
          <a
            class="card__desc"
            href="/item/4604049097548"
          >
            <span
              class="card__title"
            >
              Neutrale
            </span>
             
            гель для стирки детского белья гипоаллергенный
          </a>
          <ul
            class="card__details"
          >
            <li
              class="card__details_item"
            >
              Штрихкод:
              <span>
                4604049097548
              </span>
            </li>
            <li
              class="card__details_item"
            >
              Производитель:
              <span>
                Neutrale
              </span>
            </li>
            <li
              class="card__details_item"
            >
              Бренд:
              <span>
                Neutrale
              </span>
            </li>
            <li
              class="card__details_item"
            >
              Тип ухода:
              <span />
            </li>
          </ul>
          <div
            class="card__payload"
          >
            <span
              class="card__cost"
            >
              281
               ₸
            </span>
            <button
              class="card__add"
            >
              В КОРЗИНУ
            </button>
            <span
              class="card__add_icon"
            />
          </div>
        </li>
      </ul>
      <div
        class="pagination-block"
      >
        <div
          class="pagination"
        >
          <button
            class="pagination__prev"
          >
            <img
              alt="prev"
              src="pagination.svg"
            />
          </button>
          <ul
            class="pagination__list"
          >
            <li
              class="pagination__item"
            >
              <button
                class="pagination__page active"
              >
                1
              </button>
            </li>
          </ul>
          <button
            class="pagination__next"
          >
            <img
              alt="next"
              src="pagination.svg"
            />
          </button>
        </div>
      </div>
      <p
        class="decription"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nullam interdum ut justo, vestibulum sagittis iaculis iaculis. Quis mattis vulputate feugiat massa vestibulum duis. Faucibus consectetur aliquet sed pellentesque consequat consectetur congue mauris venenatis. Nunc elit, dignissim sed nulla ullamcorper enim, malesuada.
      </p>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
